### ISSUE TITLE
With docker: resume training fails with: "File exists: '.../decoder'

### ISSUE THREAD

**Issue by magali-joine** [modernmt/modernmt#505]
```
./mmt create --resume en chr corpus/ --count-threshold 0 --voc-size 65536

=========== TRAINING STARTED ===========

ENGINE:   default
LANGUAGE: en > chr

WARNING: No GPU for Neural engine training, the process will take very long time to complete.

(1/4) Cleaning corpora
(1/4) Cleaning corpora SKIPPED
(2/4) Creating engine

ERROR Unexpected exception: [Errno 17] File exists: '/opt/mmt/engines/default/models/decoder'
```

**Comment by nicolabertoldi** [comment]
@magali-joine 

in the current version, we support only ISO-639-1 (two-letters) for the  the language codes.

Do you really intend to use the code for Cherokee (chr)?
If so, you can use an artificial non existing code, like "xx" instead of "chr" (Cherokee).
Remember to properly change the suffix of your training files, and the language code inside the tmx if you are exploiting that format type.

We are working just now to extend the support to ISO-639-2 (three-letters).
Stay tuned to know when this enhancement will be ready!

When the extension to ISO-639-2 will be deployed, you will not lose your trained system. You will just need few simple changes in your model setting.

**Comment by magali-joine** [comment]
Thanks. Will test. (Only have CPU for modelling... so not exactly fast here).

BTW: What is a good "loss" number to expect at least partially correct results?

**Comment by magali-joine** [comment]
Does the following seem as a valid work around for the cross device link issue?

```bash
#!/bin/bash

SRC="en"
TGT="xx"

cd "$(dirname "$0")"

BASE="$(pwd)"

DATA="$BASE/corpus.${SRC}_${TGT}"
ENGINE="$BASE/engine.${SRC}_${TGT}"

if [ ! -d "$DATA" ]; then mkdir -p "$DATA"; fi
if [ ! -d "$ENGINE" ]; then mkdir -p "$ENGINE"; fi

echo "DATA: $DATA"
echo "ENGINE: $ENGINE"

echo "#!/bin/bash" > "$ENGINE/train-engine.sh"
echo "cd /opt/mmt" >> "$ENGINE/train-engine.sh"
echo "if [ ! -d engines/runtime.${SRC}_${TGT} ]; then mkdir engines/runtime.${SRC}_${TGT}; fi" >> "$ENGINE/train-engine.sh"
echo "if [ -d runtime ]; then rmdir runtime; fi" >> "$ENGINE/train-engine.sh"
echo "ln -sf engines/runtime.${SRC}_${TGT} runtime" >> "$ENGINE/train-engine.sh"
echo "./mmt create --voc-size 65536 --count-threshold 0 en xx corpus" >> "$ENGINE/train-engine.sh"
echo "/bin/bash" >> "$ENGINE/train-engine.sh"

chmod +x  "$ENGINE/train-engine.sh"

#
docker run -it \
    --mount type=bind,src="$ENGINE",target="/opt/mmt/engines" \
    --mount type=bind,src="$DATA",target="/opt/mmt/corpus" \
    modernmt/master "/opt/mmt/engines/train-engine.sh" #bash #
```

which generates the file:

```bash
#!/bin/bash
cd /opt/mmt
if [ ! -d engines/runtime.en_xx ]; then mkdir engines/runtime.en_xx; fi
if [ -d runtime ]; then rmdir runtime; fi
ln -sf engines/runtime.en_xx runtime
./mmt create --voc-size 65536 --count-threshold 0 en xx corpus
/bin/bash
```

**Comment by nicolabertoldi** [comment]
@magali-joine 

very hard to say
because it depends on several aspects:
- amount and variability of training data
- language difficulty
- distance between source and target languages,
- etc.

You can monitor the plot of the loss (and other metrics) using tensorboard.
Start the tensorboards server as follows:
```
tensorboard --logdir /opt/mmt/runtime/default/tmp/training/_temp_train/tensorboard_logdir/
```
and then connect to the server from a browser at this assuming the server is running on the localhost
```
address http://192.0.0.1:6006
```

When the loss curve is "flat" enough, you could try start evaluation.

**Comment by magali-joine** [comment]
After changing to 'xx' instead of 'chr', the training resume still fails with the 'decoder' file exists error.

```bash
muksihs@Precision-Workstation:~/modern-mt$ ./train-en-chr.sh
DATA: /home/muksihs/modern-mt/corpus.en_xx
ENGINE: /home/muksihs/modern-mt/engine.en_xx

=========== TRAINING STARTED ===========

ENGINE:   default
LANGUAGE: en > xx

WARNING: No GPU for Neural engine training, the process will take very long time to complete.

(1/4) Cleaning corpora
    (1/6) Cleaning corpora...                                    DONE in 6s
    (2/6) De-duplicate corpora...                                DONE in 1s
    (3/6) Preprocess corpora...                                  DONE in 5s
    (4/6) Creating aligner filter...                             DONE in 16s
    (5/6) Scoring corpora...                                     DONE in 3s
    (6/6) Apply aligner-based filter...                          DONE in 0s
(1/4) Cleaning corpora DONE in 32s
(2/4) Creating engine
(2/4) Creating engine DONE in 0s
(3/4) Generating binary archives
    (1/4) Tokenize corpora...                                    DONE in 5s
    (2/4) Training BPE model...                                  DONE in 1m 46s
    (3/4) Encoding training corpora...                           DONE in 2s
    (4/4) Generating binary data...                              DONE in 9s
(3/4) Generating binary archives DONE in 2m 3s
(4/4) Training neural model
    (1/2) Train neural network...                                ^CDONE in 11h 13m 10s
    (2/2) Averaging checkpoints...                               DONE in 1s
(4/4) Training neural model DONE in 11h 13m 12s

=========== TRAINING SUCCESS ===========

You can now start, stop or check the status of the server with command:
        ./mmt start|stop|status 

root@d03719e23e7a:/opt/mmt# ./mmt create --resume --voc-size 65536 --count-threshold 0 en xx corpus

=========== TRAINING STARTED ===========

ENGINE:   default
LANGUAGE: en > xx

WARNING: No GPU for Neural engine training, the process will take very long time to complete.

(1/4) Cleaning corpora
    (1/6) Cleaning corpora...                                    DONE in 6s
    (2/6) De-duplicate corpora...                                DONE in 1s
    (3/6) Preprocess corpora...                                  DONE in 5s
    (4/6) Creating aligner filter...                             DONE in 15s
    (5/6) Scoring corpora...                                     DONE in 2s
    (6/6) Apply aligner-based filter...                          DONE in 0s
(1/4) Cleaning corpora DONE in 31s
(2/4) Creating engine

ERROR Unexpected exception: [Errno 17] File exists: '/opt/mmt/engines/default/models/decoder'

```

**Comment by nicolabertoldi** [comment]
@magali-joine 
I cannot replicate this problem
Problem a previous incomplete training leaves some unexpected wrong files.

Which is the content of '/opt/mmt/engines/default/models/decoder' and its subdirectories?
Which version of ModernMT are you using?

Could you please:
- remove the model and runtime directories of you engine 
```
/opt/mmt/engines/default/
/opt/mmt/runtime/default/
```
- run the following   command
```
./mmt create --resume --voc-size 65536 --count-threshold 0 en xx corpus --debug
```
and in case of error, please report the content of log in 
```/opt/mmt/runtime/default/logs/training.log```

**Comment by magali-joine** [comment]
### File listing:
```
muksihs@Precision-Workstation:~/modern-mt/engine.en_xx/default$ tree
.
├── engine.xconf
├── models
│   └── decoder
│       ├── en__xx
│       │   ├── model.pt
│       │   └── model.vcb
│       └── model.conf
└── test_data
    └── en__xx
        ├── ced.en
        ├── ced.xx
        ├── corpus-genesis-asv.en
        ├── corpus-genesis-asv.xx
        ├── corpus-genesis-bbe.en
        ├── corpus-genesis-bbe.xx
        ├── corpus-genesis-dar.en
        ├── corpus-genesis-dar.xx
        ├── corpus-genesis-wbt.en
        ├── corpus-genesis-wbt.xx
        ├── corpus-genesis-web.en
        ├── corpus-genesis-web.xx
        ├── corpus-genesis-ylt.en
        ├── corpus-genesis-ylt.xx
        ├── corpus-nt-asv.en
        ├── corpus-nt-asv.xx
        ├── corpus-nt-bbe.en
        ├── corpus-nt-bbe.xx
        ├── corpus-nt-dar.en
        ├── corpus-nt-dar.xx
        ├── corpus-nt-wbt.en
        ├── corpus-nt-wbt.xx
        ├── corpus-nt-web.en
        ├── corpus-nt-web.xx
        ├── corpus-nt-ylt.en
        ├── corpus-nt-ylt.xx
        ├── corpus-synthetic.en
        ├── corpus-synthetic.xx
        ├── foxy-fagan.en
        ├── foxy-fagan.xx
        ├── jitaga-usdi.en
        ├── jitaga-usdi.xx
        ├── phoenix.en
        ├── phoenix.xx
        ├── three-kittens.en
        └── three-kittens.xx

5 directories, 40 files
```

### mmt version
* docker image "modernmt/master"

### verbose rm of /opt/mmt/engines/default/ /opt/mmt/runtime/default/
```
root@0f70a0e60dd2:/opt/mmt# rm -rfv /opt/mmt/engines/default/ /opt/mmt/runtime/default/
removed '/opt/mmt/engines/default/engine.xconf'
removed '/opt/mmt/engines/default/test_data/en__xx/foxy-fagan.en'
removed '/opt/mmt/engines/default/test_data/en__xx/corpus-genesis-bbe.en'
removed '/opt/mmt/engines/default/test_data/en__xx/corpus-genesis-web.xx'
removed '/opt/mmt/engines/default/test_data/en__xx/corpus-nt-bbe.en'
...
removed '/opt/mmt/runtime/default/tmp/training/corpora_clean/corpus-genesis-web.en'
removed directory '/opt/mmt/runtime/default/tmp/training/corpora_clean'
removed '/opt/mmt/runtime/default/tmp/training/state.json'
removed directory '/opt/mmt/runtime/default/tmp/training'
removed directory '/opt/mmt/runtime/default/tmp'
removed directory '/opt/mmt/runtime/default/'
```

### training with --debug results
```
muksihs@Precision-Workstation:~/modern-mt$ ./train-en-chr.sh 
DATA: /home/muksihs/modern-mt/corpus.en_xx
ENGINE: /home/muksihs/modern-mt/engine.en_xx

=========== TRAINING STARTED ===========

ENGINE:   default
LANGUAGE: en > xx

WARNING: No GPU for Neural engine training, the process will take very long time to complete.

(1/4) Cleaning corpora
    (1/6) Cleaning corpora...                                    DONE in 9s
    (2/6) De-duplicate corpora...                                DONE in 1s
    (3/6) Preprocess corpora...                                  DONE in 9s
    (4/6) Creating aligner filter...                             DONE in 29s
    (5/6) Scoring corpora...                                     DONE in 5s
    (6/6) Apply aligner-based filter...                          DONE in 0s
(1/4) Cleaning corpora DONE in 56s
(2/4) Creating engine

ERROR Unexpected exception: [Errno 2] No such file or directory: '/opt/mmt/engines/default/engine.xconf'
root@98d7ef324aa7:/opt/mmt# 
```

### training log file:
[training.log](https://github.com/modernmt/modernmt/files/3600632/training.log)

**Comment by magali-joine** [comment]
Just in case it might remotely be useful:

### corpus data
[Cherokee-English Corpus Data](https://github.com/CherokeeLanguage/CherokeeEnglishCorpus/tree/master/corpus.aligned)

**Comment by magali-joine** [comment]
Figured I'd try with a persistent volume instead of a local file system bind, and I get the same error on resume.

When I hit ctrl-c in docker, it seems to lose the state.json file after averaging checkpoints into a model? (and re-creates it when trying to resume?)

```bash
root@8510072896af:/opt/mmt# find . -name 'state.json'
./runtime/default/tmp/training/state.json
```

```bash
root@8510072896af:/opt/mmt# cat ./runtime/default/tmp/training/state.json
```
```json
{
  "corpora_clean": "/opt/mmt/runtime/default/tmp/training/corpora_clean",
  "step_no": 0
}
```

**Comment by nicolabertoldi** [comment]
@magali-joine 

if you do NOT use `--debug`
the temporary training directory is removed when the training is completed or interrupted by signal '-2' (CTRL-C)

As you already understood, the temporary training directory is kept if the debugging mode is active.

Feel free to close the issue, if you are ok.
